### ISSUE TITLE
Add declaredLicenseOf relationship

### ISSUE THREAD

**Issue by JPEWdev** [spdx/spdx-3-model#543]
Currently, the declaredLicense relationship has high fan out from the artifact to multiple licenses. While this is probably useful for large combined artifacts where multiple licenses are possible, there is another case that is less than ideal. When scanning source files for SPDX identifiers in a given project, it is much more likely that many of the files will have the same license in them, but this currently requires a relationship for each source file. It would be useful to add a new relationship that goes `from` a (single) license `to` multiple software artifacts, possibly named `declaredLicenseOf`

**Comment by goneall** [comment]
I agree this would be more compact and easier to generate, I struggle a bit with the semantics of this direction of a license relationship.  We don't want to imply that "here's all the artifacts the are using the license".  Of course, we could cover this using the completeness property, I would worry a lot of people would just leave the defaults.

BTW - I don't feel strongly about this, just pointing out a possible confusing interpretation.

**Comment by zvr** [comment]
We probably have different experiences, but this proposal will result into something much more difficult to generate and process! Compare:

**Current status: Artifact ---declareLicense--> License**

- for each file:
  - scan the file `fname`, find the declared license `X`, generate the relationship `fname declaredLicense X`.

Keep in mind that there is no fan-out, but a single license `X` -- which can be a license expression (like `BSD-3-Clause OR MIT`). The licensing profile conformance actually specifies this.

**Proposed: License ---declaredLicenseOf--> Artifact**

- for each file:
  - scan the file `fname`, find the declared license `X`, remember the association between them
- after you've finished scanning, for each of found license expressions `x`:
  - for each of the files `f_i` associated with `x`, add it to a list `flist`
  - generate the relationship `x declaredLicenseOf flist`

Essentially, you need to keep all licensing information in the project in memory (or somewhere) before generating relationships describing sets of files. As an example, the current Linux kernel has 66556 files under 116 different license expressions (e.g., 1295 under `MIT`, 16653 under `GPL-2.0-only`, etc.). The proposed design would necessitate keeping the licensing info of the 66556 files (ideally indexed by license, so using a 116-element array) before generating the relationships. I find this much more difficult to generate.

And then let's talk about having to update a finding: a file was initially marked `GPL-2.0-only` but upon further scrutiny, it was decided the declared license should be ` GPL-2.0-or-later`.
In the current state, we can easily see how this is accomplished, either by having a new relationship or amending the current one.
In the proposed design, one will have to have a new relationship for the 16652 (= 16653-1) files that are under `GPL-2.0-only` and a new one for 8461 (= 1+8460 files that were previously under `GPL-2.0-or-later`) ones.

Not to mention the case when someone wants to add a comment (an Annotation) to say "this file only says `GPL`, so we use `GPL-1.0-or-later`".

In conclusion I am strongly against reversing the direction of licensing relationships.
Going from license to artifact is the equivalent of replacing the `createdBy` property by `creatorOf`. Yes, the same information can be expressed in either form, but one is much, much more easier to handle.

**Comment by JPEWdev** [comment]
Yes it can work either way (although, I would recommend better relationship names in general, per #534). Just because the relationship exists with high fan out doesn't mean a given implementation _has_ to group them that way. Even if the `declaredLicenseOf` existed, implementations would not be required to use the high fan out if memory is a concern. It's possible to still do the 1:1 mapping, as you are basically forced to do with the existing implementation. However, if you're willing to take the overhead hit, the high fan out gives the option to group them together.

FWIW, I don't really care too much about this; It's not hard to programmatically generated a ton of relationships; I care a lot more about #534 and #533 because it's basically impossible to figure out how to correctly use relationships currently. However, given that we (The Yocto Project) can generate such a large volume of SPDX data, I was encouraged to raise issues for places where the fan-out of relationships was not ideal (e.g. where we had to make a lot of relationships vs. one with a larger fan out, since the latter results in smaller file sizes).

**Comment by goneall** [comment]
@kestewart - Is this resolved in one of you pull requests?

**Comment by sbarnum** [comment]
> We probably have different experiences, but this proposal will result into something much more difficult to generate and process! Compare:
> 
> **Current status: Artifact ---declareLicense--> License**
> 
> * for each file:
>   
>   * scan the file `fname`, find the declared license `X`, generate the relationship `fname declaredLicense X`.
> 
> Keep in mind that there is no fan-out, but a single license `X` -- which can be a license expression (like `BSD-3-Clause OR MIT`). The licensing profile conformance actually specifies this.
> 
> **Proposed: License ---declaredLicenseOf--> Artifact**
> 
> * for each file:
>   
>   * scan the file `fname`, find the declared license `X`, remember the association between them
> * after you've finished scanning, for each of found license expressions `x`:
>   
>   * for each of the files `f_i` associated with `x`, add it to a list `flist`
>   * generate the relationship `x declaredLicenseOf flist`
> 
> Essentially, you need to keep all licensing information in the project in memory (or somewhere) before generating relationships describing sets of files. As an example, the current Linux kernel has 66556 files under 116 different license expressions (e.g., 1295 under `MIT`, 16653 under `GPL-2.0-only`, etc.). The proposed design would necessitate keeping the licensing info of the 66556 files (ideally indexed by license, so using a 116-element array) before generating the relationships. I find this much more difficult to generate.
> 
> And then let's talk about having to update a finding: a file was initially marked `GPL-2.0-only` but upon further scrutiny, it was decided the declared license should be ` GPL-2.0-or-later`. In the current state, we can easily see how this is accomplished, either by having a new relationship or amending the current one. In the proposed design, one will have to have a new relationship for the 16652 (= 16653-1) files that are under `GPL-2.0-only` and a new one for 8461 (= 1+8460 files that were previously under `GPL-2.0-or-later`) ones.
> 
> Not to mention the case when someone wants to add a comment (an Annotation) to say "this file only says `GPL`, so we use `GPL-1.0-or-later`".
> 
> In conclusion I am strongly against reversing the direction of licensing relationships. Going from license to artifact is the equivalent of replacing the `createdBy` property by `creatorOf`. Yes, the same information can be expressed in either form, but one is much, much more easier to handle.

I would agree with this.
While declaredLicenseOf could potentially yield more terse graph content I believe it would likely add significantly more headaches in processing, parsing, understanding, and maintaining such graphs.
I would vote against this approach.

**Comment by kestewart** [comment]
In https://github.com/spdx/spdx-3-model/blob/main/model/Core/Vocabularies/RelationshipType.md
we have 
"hasDeclaredLicense: The from Software Artifact was discovered to actually contain each to license, for example as detected by use of automated tooling."
